runtime/process.go

```diff
@@ -195,7 +229,47 @@ func (p *process) Resize(w, h int) error {
 }
 
 func (p *process) handleSigkilledShim(rst int, rerr error) (int, error) {
-	if rerr == nil || p.cmd == nil || p.cmd.Process == nil {
+	if p.cmd == nil || p.cmd.Process == nil {
+		e := unix.Kill(p.pid, 0)
+		if e == syscall.ESRCH {
+			return rst, rerr
+		}
+
+		// If it's not the same process, just mark it stopped and set
+		// the status to 255
+		if same, err := p.isSameProcess(); !same {
+			logrus.Warnf("containerd: %s:%s (pid %d) is not the same process anymore (%v)", p.container.id, p.id, p.pid, err)
+			p.stateLock.Lock()
+			p.state = Stopped
+			p.stateLock.Unlock()
+			// Create the file so we get the exit event generated once monitor kicks in
+			// without going to this all process again
+			rerr = ioutil.WriteFile(filepath.Join(p.root, ExitStatusFile), []byte("255"), 0644)
```

Comment:
because we set the parent death signal for the shim, we can assume that the container process also gets a SIGKILL when the shim dies.